src/group_theory/monoid_localization.lean

```diff
@@ -6,410 +6,316 @@ Authors: Amelia Livingston
+predicate.
 
 ## Implementation notes
 
-The infimum form of the localization congruence relation is chosen as 'canonical' here, since it
-shortens many proofs.
+In maths it is natural to reason up to isomorphism, but in Lean we cannot naturally `rewrite` one
+structure with an isomorphic one; one way around this is to isolate a predicate characterizing
+a structure up to isomorphism, and reason about things that satisfy the predicate.
 
-The `private def` `r'.rel` and the lemmas `r'.add, r'.transitive` are to enable the use of the
-`abel` tactic for both the additive and multiplicative proofs that the 'usual' localization
-congruence relation is a congruence relation.
+The infimum form of the localization congruence relation is chosen as 'canonical' here, since it
+shortens some proofs.
 
-There is only a multiplicative version for any lemma or definition relying on a unit group of a
-`comm_monoid`; additive versions would require additive unit groups.
+We use `classical.epsilon` where possible; like `classical.some`, it reduces the number of
+arguments to functions, but shortens proofs in comparison to `classical.some`.
 
 ## Tags
-localization, monoid localization, quotient monoid, congruence relation
-
+localization, monoid localization, quotient monoid, congruence relation, characteristic predicate,
+commutative monoid
 -/
-variables {X : Type*}
+variables {X : Type*} [comm_monoid X] (Y : submonoid X) {Z : Type*} [comm_monoid Z]
+          {V : Type*} [comm_monoid V] (f : X →* Z)
 
 namespace submonoid
 
-/-- The congruence relation on `X × Y`, `X` a `comm_monoid` and `Y` a submonoid of `X`, whose
-    quotient is the localization of `X` at `Y`, defined as the unique congruence relation on
-    `X × Y` such that for any other congruence relation `s` on `X × Y` where for all `y ∈ Y`,
-    `(1, 1) ∼ (y, y)` under `s`, we have that `(x₁, y₁) ∼ (x₂, y₂)` by `s` implies
-    `(x₁, y₁) ∼ (x₂, y₂)` by `r`. -/
-@[to_additive "The congruence relation on `X × Y`, `X` an `add_comm_monoid` and `Y` an `add_submonoid` of `X`, whose quotient is the localization of `X` at `Y`, defined as the unique congruence relation on `X × Y` such that for any other congruence relation `s` on `X × Y` where for all `y ∈ Y`, `(0, 0) ∼ (y, y)` under `s`, we have that `(x₁, y₁) ∼ (x₂, y₂)` by `s` implies `(x₁, y₁) ∼ (x₂, y₂)` by `r`."]
-def r [comm_monoid X] (Y : submonoid X) : con (X × Y) :=
-lattice.Inf {c | ∀ y : Y, c 1 (y, y)}
-
-end submonoid
-
-namespace add_submonoid
-
-variables [add_comm_monoid X]
+/-- The characteristic predicate: if `f` satisfies `is_localization_map`, then `Z` is isomorphic
+    to the localization of `X` at `Y`. -/
+@[to_additive "The characteristic predicate: if `f` satisfies `is_localization_map`, then `Z` is isomorphic to the localization of `X` at `Y`."]
+def is_localization_map :=
+(∀ y : Y, ∃ z, f y * z = 1) ∧ (∀ z : Z, ∃ x : X × Y, z * f x.2 = f x.1) ∧
```

Comment:
Why not use `is_unit (f y)`?

Reply:
I'd use `structure` with descriptive field names here.

Reply:
Thanks for the review! Sorry for the slow response. One reason is ```units``` stuff hasn't been ```to_additive```'d. A less good reason is the ```units``` structure contains redundant information in a ```comm_monoid```, but retrospectively that's not really a reason. I'm currently refactoring to use ```is_unit``` to see if I had any other reasons.

Reply:
In the latest commit I haven't switched to ```is_unit``` yet but having tried it out, I can see it'll be an improvement to switch, if we don't care about having additive versions. 

Also, I haven't added a ```has_coe_to_fun``` for the structure ```localization_map``` yet; it ends up making proofs annoyingly longer. The first field of a ```localization_map``` ```f``` is a monoid hom and life seems to be easier just using ```f.1``` everywhere.

Reply:
I'm a bit on the fence on this one... because I can imagine that `nat → int` is a nice example of an additive localization map.
Maybe we should just bite the bullet and add `to_additive` tags to all the units stuff? @urkud what do you think?

Reply:
I think it's OK to add `to_additive` to all `units` / `is_unit` lemmas in `algebra/group`. How would you name it for additive monoids?

Reply:
It's ugly, but `add_units` is the first thing that comes to mind.

Reply:
Is it alright if I commit some (minimal) ```add_units``` stuff on this PR, rather than making a separate PR?

Reply:
That's fine with me.